### PR TITLE
EDUCATOR-1787 Allow hyphen character in course number

### PR DESCRIPTION
**Pull request by Rabia23**
### Description
In the past, publisher only allows [ _ ] and [ . ] (In addition to alphanumeric characters) in course number but now it allows [-] character as well to keep it consistent with Studio which allows  [ _ ], [ - ], [ . ]. It also allows [.] in organization key (In addition to alphanumeric characters, underscore and hyphen) for consistency with studio as we had missed this case in https://github.com/edx/course-discovery/pull/1203.

### How to Test?
**Sandbox**
https://discovery-escalation-discovery.sandbox.edx.org/publisher/courses/29/edit/
Edit any field and click on `Update Course`. It allows [-] character in course number.

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x]  @awaisdar001 
- [x]  @asadazam93 

FYI: 

### Post-review
- [x] Rebase and squash commits

### CODE REVIEW

File: course_discovery/apps/publisher/templates/publisher/course_edit_form.html [hunk 1]
```diff
@@ -131,7 +131,7 @@ <h1 class="hd-1 emphasized">{% trans "Edit Course" %}</h1>
                                         </div>
                                         <div id="tab-practices" class="content active">
                                             <ul>
-                                                <li>{% trans "Maximum 10 characters. Characters can be letters, numbers, or periods." %}</li>
+                                                <li>{% trans "Maximum 50 characters. Characters can be letters, numbers, periods, underscores or hyphens." %}</li>
```
Comment (awaisdar001) [inline course_discovery/apps/publisher/templates/publisher/course_edit_form.html:134]: is it just the cosmetic change? changing 10 to 50?

File: course_discovery/apps/publisher/tests/test_forms.py [hunk 2]
```diff
@@ -223,21 +225,36 @@ def test_duplicate_number(self):
         course_form.cleaned_data['number'] = "123a"
         self.assertEqual(course_form.clean(), course_form.cleaned_data)
 
-    def test_invalid_number(self):
+    @ddt.data(
+        "123 a",
+        "abc@123",
+        "abc(123)"
```
Comment (awaisdar001) [inline course_discovery/apps/publisher/tests/test_forms.py]: also check for `!#$%^&*+=` chars in the dd.data.

Comment (awaisdar001) [reply course_discovery/apps/publisher/tests/test_forms.py]: one thing you can do is to pass the invalid chars as a list in dd.data and on line 239 you can combine the chars in iteration with and test the validation error.

**Review by awaisdar001 — CHANGES_REQUESTED** [review]

File: course_discovery/apps/publisher/templates/publisher/course_edit_form.html (hunk 1 above)
Comment (Rabia23) [reply course_discovery/apps/publisher/templates/publisher/course_edit_form.html:134]: yes

File: course_discovery/apps/publisher/tests/test_forms.py (hunk 2 above)
Comment (Rabia23) [reply course_discovery/apps/publisher/tests/test_forms.py]: sure

**Comment by Rabia23** [comment]
@awaisdar001 I have addressed your changes. Can you please take a look again?

File: course_discovery/apps/publisher/tests/test_forms.py [hunk 3]
```diff
@@ -223,21 +225,35 @@ def test_duplicate_number(self):
         course_form.cleaned_data['number'] = "123a"
         self.assertEqual(course_form.clean(), course_form.cleaned_data)
 
-    def test_invalid_number(self):
+    @ddt.data(
+        [" ", "@", "(", "!", "#", "$", "%", "^", "&", "*", "+", "=", "{", "["]
+    )
+    def test_invalid_course_number(self, number_list):
```
Comment (awaisdar001) [inline course_discovery/apps/publisher/tests/test_forms.py]: `invalid_char_list`

File: course_discovery/apps/publisher/tests/test_forms.py [hunk 4]
```diff
@@ -223,21 +225,35 @@ def test_duplicate_number(self):
...
+                course_form.clean_number()
+
+    @ddt.data(
+        "123a",
+        "123_a",
+        "123.a",
+        "123-a",
+        "XYZ123"
+    )
+    def test_valid_course_number(self, number):
```
Comment (awaisdar001) [inline course_discovery/apps/publisher/tests/test_forms.py]: `number` => `valid_number`

File: course_discovery/apps/publisher/templates/publisher/course_edit_form.html [hunk 5]
```diff
@@ -145,6 +145,11 @@ <h1 class="hd-1 emphasized">{% trans "Edit Course" %}</h1>
                                     <div class="col col-6">
                                         <label class="field-label ">{{ form.number.label_tag }} <span class="required">*</span></label>
                                         {{ form.number }}
+                                        {% if form.number.errors %}
```
Comment (awaisdar001) [inline course_discovery/apps/publisher/templates/publisher/course_edit_form.html:148]: Why are we using `escape`?

File: course_discovery/apps/publisher/forms.py [hunk 6]
```diff
@@ -182,11 +182,12 @@ def clean_title(self):
 
     def clean_number(self):
         """
-        Validate that number doesn't consist of any special characters
+        Validate that number doesn't consist of any special characters other than period, underscore or hyphen
         """
         number = self.cleaned_data.get("number")
-        if not re.match("^[a-zA-Z0-9_.]*$", number):
-            raise ValidationError(_('Please do not use any spaces or special characters.'))
+        if not re.match("^[a-zA-Z0-9._-]*$", number):
```
Comment (awaisdar001) [inline course_discovery/apps/publisher/forms.py]: Can publisher and course metadata both use the same regex?

**Review by awaisdar001 — CHANGES_REQUESTED** [review]

File: course_discovery/apps/course_metadata/models.py [hunk 7]
```diff
@@ -176,7 +177,7 @@ class Organization(TimeStampedModel):
     """ Organization model. """
     partner = models.ForeignKey(Partner, null=True, blank=False)
     uuid = models.UUIDField(blank=False, null=False, default=uuid4, editable=False, verbose_name=_('UUID'))
-    key = models.CharField(max_length=255, help_text=_('Only ascii characters allowed (a-zA-Z0-9)'))
+    key = models.CharField(max_length=255, help_text=_('Only ascii characters allowed (a-zA-Z0-9._-)'))
```
Comment (awaisdar001) [inline course_discovery/apps/course_metadata/models.py]: @Rabia23  Can we have the `help_text` more like the validation error to be more explicit about what characters we are allowing to course teams?

File: course_discovery/apps/publisher/utils.py [hunk 8]
```diff
@@ -1,10 +1,12 @@
 """ Publisher Utils."""
+import re
 from dateutil import parser
 
 from course_discovery.apps.core.models import User
 from course_discovery.apps.publisher.constants import (ADMIN_GROUP_NAME, INTERNAL_USER_GROUP_NAME,
                                                        PROJECT_COORDINATOR_GROUP_NAME)
 
+ALLOWED_CHARS_PATTERN = re.compile(r"^[a-zA-Z0-9._-]*$")
```
Comment (awaisdar001) [inline course_discovery/apps/publisher/utils.py]: Which chars? The variable is not self-explanatory.

File: course_discovery/apps/course_metadata/tests/test_models.py [hunk 9]
```diff
@@ -366,21 +366,21 @@ def setUp(self):
         self.organization = factories.OrganizationFactory()
 
     @ddt.data(
-        "key with space",
-        "key[with,special",
-        "keyó"
+        [" ", "@", "(", "!", "#", "$", "%", "^", "&", "*", "+", "=", "{", "[", "ó"]
```
Comment (awaisdar001) [inline course_discovery/apps/course_metadata/tests/test_models.py]: Do we need to test `,` also?

File: course_discovery/apps/course_metadata/tests/test_models.py [hunk 10]
```diff
@@ -366,21 +366,21 @@ def setUp(self):
...
 
     @ddt.data(
         "keywithoutspace",
-        "correctkey",
-        "correct_key"
+        "correct-key",
+        "correct_key",
+        "correct.key"
     )
     def test_clean_success(self, key):
```
Comment (awaisdar001) [inline course_discovery/apps/course_metadata/tests/test_models.py]: what is `key` here? we are not unpacking ddt.data here. is the variable `key` is a list/tuple?

File: course_discovery/apps/publisher/templates/publisher/course_edit_form.html (hunk 5 above)
Comment (awaisdar001) [reply course_discovery/apps/publisher/templates/publisher/course_edit_form.html:148]: @Rabia23 ^^

File: course_discovery/apps/publisher/tests/test_forms.py [hunk 11]
```diff
@@ -223,21 +225,35 @@ def test_duplicate_number(self):
...
+                course_form.clean_number()
+
+    @ddt.data(
+        "123a",
+        "123_a",
+        "123.a",
+        "123-a",
+        "XYZ123"
+    )
+    def test_valid_course_number(self, valid_number):
```
Comment (awaisdar001) [inline course_discovery/apps/publisher/tests/test_forms.py]: See my comment above.
`valid_number` same as `key`

File: course_discovery/apps/publisher/forms.py [hunk 12]
```diff
@@ -182,11 +181,12 @@ def clean_title(self):
...
     def clean_number(self):
         """
-        Validate that number doesn't consist of any special characters
+        Validate that number doesn't consist of any special characters other than period, underscore or hyphen
         """
         number = self.cleaned_data.get("number")
-        if not re.match("^[a-zA-Z0-9_.]*$", number):
-            raise ValidationError(_('Please do not use any spaces or special characters.'))
+        if not ALLOWED_CHARS_PATTERN.match(number):
+            raise ValidationError(_('Please do not use any spaces or special characters other than period, '
```
Comment (awaisdar001) [inline course_discovery/apps/publisher/forms.py]: @Rabia23 Please find a perfered way to use string in edX. Here you are using `''` to define a string and on the L#186 it `self.cleaned_data.get("number")` please find a perfered way and stick with it.

**Review by awaisdar001 — CHANGES_REQUESTED** [review]

File: course_discovery/apps/course_metadata/tests/test_models.py (hunk 10 above)
Comment (Rabia23) [reply course_discovery/apps/course_metadata/tests/test_models.py]: `key` is a variable name. Each value will be assigned to variable `key`.

Comment (awaisdar001) [reply course_discovery/apps/course_metadata/tests/test_models.py]: see no `@ddt.unpack`?

Comment (Rabia23) [reply course_discovery/apps/course_metadata/tests/test_models.py]: because there is no need of it.

Comment (Rabia23) [reply course_discovery/apps/course_metadata/tests/test_models.py]: `@ddt.data(
        "keywithoutspace",
        "correct-key",
        "correct_key",
        "correct.key"
    )`
These are four individual tests.

Comment (awaisdar001) [reply course_discovery/apps/course_metadata/tests/test_models.py]: you are assigning the list to `self.organization.key = key` in `test_clean_success` body and then calling a clean, are you missing something?

**Review by awaisdar001 — CHANGES_REQUESTED** [review]

File: course_discovery/apps/course_metadata/tests/test_models.py (hunk 10 above)
Comment (Rabia23) [reply course_discovery/apps/course_metadata/tests/test_models.py]: I am not assigning list to `self.organization.key`. Individual value is assigned to `self.organization.key`.

Comment (awaisdar001) [reply course_discovery/apps/course_metadata/tests/test_models.py]: Talked with rabia. we are on the same page with this one. Got rid of my confusion on this.

File: course_discovery/apps/publisher/tests/test_forms.py (hunk 11 above)
Comment (Rabia23) [reply course_discovery/apps/publisher/tests/test_forms.py]: we are on the same page for `valid_number` as we discussed above for `key` variable.

**Comment by awaisdar001** [comment]
@Rabia23 please read out https://openedx.atlassian.net/wiki/spaces/OpenDev/pages/161880915/Git+Best+Practices and do not squash commits once a review has started, to aid reviewers.

File: course_discovery/apps/course_metadata/models.py [hunk 13]
```diff
@@ -176,7 +177,9 @@ class Organization(TimeStampedModel):
     """ Organization model. """
     partner = models.ForeignKey(Partner, null=True, blank=False)
     uuid = models.UUIDField(blank=False, null=False, default=uuid4, editable=False, verbose_name=_('UUID'))
-    key = models.CharField(max_length=255, help_text=_('Only ascii characters allowed (a-zA-Z0-9)'))
+    key = models.CharField(max_length=255, help_text=_('Please do not use any spaces or special characters other '
+                                                       'than period, underscore or hyphen. This key will be used '
+                                                       'in the course\'s course key.'))
```
Comment (awaisdar001) [inline course_discovery/apps/course_metadata/models.py:182]: 👍

File: course_discovery/apps/publisher/tests/test_forms.py [hunk 14]
```diff
@@ -223,21 +225,35 @@ def test_duplicate_number(self):
...
-        with self.assertRaises(ValidationError):
-            course_form.clean_number()
-
-        course_form.cleaned_data['number'] = "123.a"
-        self.assertEqual(course_form.clean_number(), "123.a")
-
-        course_form.cleaned_data['number'] = "123a"
-        self.assertEqual(course_form.clean_number(), "123a")
+        for char in invalid_char_list:
+            course_form.cleaned_data = {'number': 'course_num{}'.format(char)}
```
Comment (awaisdar001) [inline course_discovery/apps/publisher/tests/test_forms.py]: `char` => `invalid_char`

**Review by awaisdar001 — APPROVED** [review]
small nits, rest looks good to me.

**Review by asadazam93 — APPROVED** [review]

**Pull request merged**